File: src/main/java/DataStructureAlgo/Java/LeetCode2025/ProblemSet/Trees/_116/PopulatingNextRightPointersInEachNode_116.java
package DataStructureAlgo.Java.LeetCode2025.ProblemSet.Trees._116;

import java.util.LinkedList;
import java.util.Queue;

/**
 * Author: Nitin Gupta
 * Date: 8/19/2024
 * Question Category: 116. Populating Next Right Pointers in Each Node
 * Description: https://leetcode.com/problems/populating-next-right-pointers-in-each-node/description/
 * You are given a perfect binary tree where all leaves are on the same level, and every parent has two children. The binary tree has the following definition:
 * <p>
 * struct Node {
 * int val;
 * Node *left;
 * Node *right;
 * Node *next;
 * }
 * Populate each next pointer to point to its next right node. If there is no next right node, the next pointer should be set to NULL.
 * <p>
 * Initially, all next pointers are set to NULL.
 * <p>
 * <p>
 * <p>
 * Example 1:
 * <p>
 * <p>
 * Input: root = [1,2,3,4,5,6,7]
 * Output: [1,#,2,3,#,4,5,6,7,#]
 * Explanation: Given the above perfect binary tree (Figure A), your function should populate each next pointer to point to its next right node, just like in Figure B. The serialized output is in level order as connected by the next pointers, with '#' signifying the end of each level.
 * Example 2:
 * <p>
 * Input: root = []
 * Output: []
 * <p>
 * <p>
 * Constraints:
 * <p>
 * The number of nodes in the tree is in the range [0, 212-1].
 * -1000 <= Node.val <= 1000
 * <p>
 * <p>
 * Follow-up:
 * <p>
 * You may only use constant extra space.
 * The recursive approach is fine. You may assume implicit stack space does not count as extra space for this problem.
 * <p>
 * File reference
 * -----------
 * Duplicate {@link DataStructureAlgo.Java.LeetCode.tree.connectNodesTree.ConnectNodesNext}
 * Similar {@link}
 * extension {@link }
 * <p>
 * Tags
 * -----
 *
 * @medium
 * @LinkedList
 * @Tree
 * @Depth-FirstSearch
 * @Breadth-FirstSearch
 * @BinaryTree <p>
 * <p>
 * Company Tags
 * -----
 * @Editorial
 */
public class PopulatingNextRightPointersInEachNode_116 {


    static class Solution {


        // Definition for a Node.
        class Node {
            public Node next;
            public int val;
            public Node left;
            public Node right;

            public Node() {

            }


            public Node(int _val, Node _left, Node _right, Node _next) {
                val = _val;
                left = _left;
                right = _right;
                next = _next;
            }
        }

        static class SolutionUsingPreOrderTraversal {

            public Node connect(Node root) {
                if (root == null)
                    return root;

                //if left node exists for this root, connect it with the right node
                if (root.left != null)
                    root.left.next = root.right;

                if (root.right != null)
                    root.right.next = (root.next == null) ? null : root.next.left;

                connect(root.left);
                connect(root.right);
                return root;
            }
        }

        static class SolutionUsingLevelOrderTraversal {

            public Node connect(Node root) {
                if (root == null)
                    return root;

                Queue<Node> queue = new LinkedList<>();
                queue.offer(root);
                queue.offer(null);

                while (!queue.isEmpty()) {

                    Node current = queue.poll();

                    if (current == null) {
                        if (!queue.isEmpty())
                            queue.offer(null);
                    } else {
                        current.next = queue.peek();


                        if (current.left != null)
                            queue.offer(current.left);

                        if (current.right != null)
                            queue.offer(current.right);

                    }
                }
                return root;
            }
        }
    }
}

File: src/main/java/DataStructureAlgo/Java/LeetCode2025/ProblemSet/Trees/_116/PopulatingNextRightPointersInEachNode_116.py
# Definition for a Node.

from collections import deque
from typing import Optional

class Node:
    def __init__(self, val: int = 0, left: Optional['Node'] = None, right: Optional['Node'] = None, next: Optional['Node'] = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next



class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root:
            return root

        queue = deque([(root, 0)])  # queue root and its level=0
        previous_node, previous_level = None, -1

        while queue:
            current_node, current_lvl = queue.popleft()

            # no more node to process
            if not current_node:
                return root

            if previous_level == current_lvl and previous_node is not None:
                previous_node.next = current_node

            # update previous node and level
            previous_node, previous_level = current_node, current_lvl

            # queue left and right with next level
            if current_node.left:
                queue.append((current_node.left, current_lvl + 1))
            if current_node.right:
                queue.append((current_node.right, current_lvl + 1))

        return root


class SolutionSimplified:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root:
            return root

        queue = deque([root])  # queue root and its level=0
        previous_node= None

        while queue:
            size = len(queue)
            previous_node= None
            
            # process all nodes in the current level
            for _ in range(size):
                current_node = queue.popleft()
            
                if previous_node:
                    previous_node.next = current_node
                
                #update previous node
                previous_node = current_node
                
                #queue left and right
                if current_node.left:
                    queue.append(current_node.left)
                if current_node.right:
                    queue.append(current_node.right)
           
        return root

class SolutionRecursive:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root:
            return root

        if root.left:
            root.left.next = root.right

        if root.right:
            root.right.next = None if not root.next else root.next.left

        self.connect(root.left)
        self.connect(root.right)
        return root

def print_levels_with_next(root):
    '''Prints each level's nodes and their next pointers.'''
    level = 0
    node = root
    while node:
        curr = node
        line = []
        while curr:
            nxt = curr.next.val if curr.next else None
            line.append(f"{curr.val}(next:{nxt})")
            curr = curr.next
        print(f"Level {level}: ", ' -> '.join(line))
        node = node.left
        level += 1

if __name__ == "__main__":
    # Build a perfect binary tree:
    #        1
    #      /   \
    #     2     3
    #    / \   / \
    #   4  5  6   7
    n4 = Node(4)
    n5 = Node(5)
    n6 = Node(6)
    n7 = Node(7)
    n2 = Node(2, n4, n5)
    n3 = Node(3, n6, n7)
    root = Node(1, n2, n3)

    print("Testing Solution (BFS):")
    sol = Solution()
    connected_root = sol.connect(root)
    print_levels_with_next(connected_root)

    # Rebuild the tree for the recursive solution (since next pointers are now set)
    n4 = Node(4)
    n5 = Node(5)
    n6 = Node(6)
    n7 = Node(7)
    n2 = Node(2, n4, n5)
    n3 = Node(3, n6, n7)
    root = Node(1, n2, n3)

    print("\nTesting SolutionRecursive:")
    sol_rec = SolutionRecursive()
    connected_root_rec = sol_rec.connect(root)
    print_levels_with_next(connected_root_rec)
